DS13/code/modules/EMH/EMH.dm

```diff
@@ -0,0 +1,309 @@
+						emh = null
+					else
+						to_chat(user, "Unable to locate an EMH on the network.")
+
+/obj/machinery/emh_emitter/Destroy()
+	if(emh)
+		var/obj/machinery/emh_emitter/saveme = locate(/obj/machinery/emh_emitter) in GLOB.machines
+		if(saveme)
+			to_chat(emh, "The emitter you were using has been destroyed, your program has been transferred to avoid corruption.")
+			saveme.activate(emh)
+		else
+			to_chat(emh, "The emitter you were using was destroyed! Your program could not be shunted to another emitter and will be shut down to avoid damage.")
+			qdel(emh)
+			emh = null
+	STOP_PROCESSING(SSmachines,src)
+	. = ..()
+
+/obj/machinery/emh_emitter/proc/remove_cooldown()
+	locked = FALSE
+
+/obj/machinery/emh_emitter/attack_hand(mob/user)
+	if(istype(user, /mob/living/carbon/human/species/holographic)) //Allow them to jump to different locs.
+		var/list/jumplocs = list()
+		for(var/obj/machinery/emh_emitter/EM in GLOB.machines)
+			if(istype(EM))
+				jumplocs += EM
+		var/A
+		A = input("Transfer program to where?", "EMH subsystem", A) as null|anything in jumplocs
+		if(!A)
+			return
+		user.say("Computer, transfer emergency medical holographic program to [get_area(A)]")
+		new /obj/effect/temp_visual/dir_setting/ninja/cloak(get_turf(user), user.dir)
+		user.forceMove(get_turf(A))
+		playsound(user.loc, 'DS13/sound/effects/holofizzle.ogg', 100, 0)
+		var/obj/machinery/emh_emitter/S = A
+		emh = null
+		S.emh = user //transfer silently.
+		return
+	if(locked)
+		to_chat(user, "Unable to comply, interface cooldown in effect.")
+		return
+	if(is_occupied()) //If we have an EMH, or there is an EMH in the world
+		to_chat(user, "There is already an EMH active, summon it?")
+		var/question = alert("Summon an EMH?",name,"yes","no")
+		locked = TRUE
+		addtimer(CALLBACK(src, .proc/remove_cooldown), 30) //Add timer to cooldown
+		if(question == "yes")
+			var/mob/living/carbon/human/species/holographic/S = locate(/mob/living/carbon/human/species/holographic) in GLOB.alive_mob_list
+			if(S)
+				var/emhconsent = alert("You have been summoned by [user] in [get_area(src)]. Transfer to their location?",name,"accept","reject")
+				if(emhconsent == "accept")
+					activate(S)
+					user.say("Computer, transfer emergency medical holographic program to [get_area(src)]")
+				else
+					to_chat(user, "The EMH has declined your summons.")
+					return ..()
+	else
+		locked = TRUE
+		addtimer(CALLBACK(src, .proc/remove_cooldown), 30) //Add timer to prevent spam clicking
+		if(emh)
+			emh.death()
+		if(user)
+			user.say("Computer, activate emergency medical hologram")
+			to_chat(user, "Attempting to activate EMH ((This requires a ghost willing to control it...))")
+		activate(null) //No EMH present, try to make one!
+
+/obj/machinery/emh_emitter/process()
+	var/area/A = get_area(src) //If someone moves us, ensure we update name.
+	name = "EMH emitter ([A])"
+	if(!emh || !is_occupied()) //If we don't have an EMH, no need to process
```

Comment:
THIS SHOULD BE FIRST FOR THE PROCESS WITH A ADDED CHECK TO CONFIRM IF ITS THE SAME AREA!